File: 13-roman-to-integer/roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        map = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000
        }

        sum = 0
        i = 0

        while i < len(s):
            curr = map[s[i]]
            if i + 1 < len(s):
                next = map[s[i + 1]]
                if curr < next:
                    sum += next - curr
                    i += 2  # Skip both characters
                    continue
            sum += curr
            i += 1

        return sum

File: 13-roman-to-integer/roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        int sum = 0;
        Map<Character, Integer> map = new HashMap<>();
        map.put('I', 1);
        map.put('V', 5);
        map.put('X', 10);
        map.put('L', 50);
        map.put('C', 100);
        map.put('D', 500);
        map.put('M', 1000);

        for (int i = 0; i < s.length(); i++) {
            int curr = map.get(s.charAt(i));
            int next = 0;

            // Safe to check i+1 only if within bounds
            if (i + 1 < s.length()) {
                next = map.get(s.charAt(i + 1));
                if (curr < next) {
                    sum += (next - curr);
                    i++; // Skip the next char because it's part of the pair
                    continue;
                }
            }

            sum += curr;
        }

        return sum;
    }
}